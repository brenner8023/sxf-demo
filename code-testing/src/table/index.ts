/**
 * Created by uedc on 2021/10/11.
 */

import TestTable from './src/Table'

export { TestTable }
