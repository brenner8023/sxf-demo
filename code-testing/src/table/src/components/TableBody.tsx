
import { defineComponent, PropType } from '@vue/composition-api';
import TableTr from './TableTr';

export default defineComponent({
  name: 'TableBody',
  props: {
    tableData: {
      type: Array as PropType<any[]>,
      default: () => [],
    },
    isTableEmpty: {
      type: Boolean,
      default: true,
    },
  },
  setup (props) {
    return () => {
      if (props.isTableEmpty) {
        return <tbody>
          <div class="sxf-empty">空状态</div>
        </tbody>;
      }

      return <tbody>
        {
          props.tableData.map(rowData => {
            return <TableTr rowData={rowData}></TableTr>
          })
        }
      </tbody>;
    };
  }
});
