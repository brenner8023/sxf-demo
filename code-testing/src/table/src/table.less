
.sxf-table-wrap {
  border: 1px solid rgba(0, 0, 0, 0.12);
  border-radius: 4px;
  overflow: auto;
}

.sxf-table {
  min-width: 100%;
  border-collapse: collapse;
  text-align: left;

  th {
    height: 52px;
    padding-left: 8px;
    padding-right: 8px;
  }
}

.sxf-table__header {
  height: 52px;
  display: flex;
  align-items: center;
}

.sxf-table__row {
  border-top: 1px solid rgba(0, 0, 0, 0.12);

  &:hover {
    background-color: rgba(0, 0, 0, 0.04);
  }

  td {
    height: 52px;
    box-sizing: border-box;
    padding-left: 8px;
    padding-right: 8px;
  }
}

.sxf-table-footer {
  padding: 8px;
}

.sxf-paging {
  display: flex;
  justify-content: end;
  align-items: center;
  margin-right: 32px;

  &__input {
    width: 40px;
    border: 1px solid green;
    margin-right: 4px;
  }

  &__arrow {
    width: 14px;
    height: 14px;
    display: inline-block;
    background-size: 14px 14px;
    background-position: center;
    cursor: pointer;

    &--left {
      margin-left: 32px;
      margin-right: 16px;
      background-image: url('../img/left.svg');
    }

    &--right {
      background-image: url('../img/right.svg');
      margin-left: 16px;
      margin-right: 32px;
    }
  }
}

.sxf-table-sorter {
  width: 14px;
  height: 14px;
  background-size: 14px;
  margin-left: 4px;

  &__asc {
    background-image: url('../img/sort-asc.svg');

    &:hover {
      cursor: pointer;
      background-image: url('../img/sort-asc-selected.svg');
    }

    &--selected {
      background-image: url('../img/sort-asc-selected.svg');
    }
  }

  &__desc {
    background-image: url('../img/sort-desc.svg');

    &:hover {
      cursor: pointer;
      background-image: url('../img/sort-desc-selected.svg');
    }

    &--selected {
      background-image: url('../img/sort-desc-selected.svg');
    }
  }
}
