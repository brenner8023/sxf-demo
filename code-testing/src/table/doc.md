# 如何使用

在demo中引入：

```ts
import { TestTable } from '../src/table';
```

在页面中使用：

```html
<TestTable></TestTable>
```
# TestTable
## TestTable 参数

| 参数 | 类型 | 默认 | 说明 |
| :-: | :-: | :-: | - |
|   data  | `any[]` | `[]` | 必选，表格数据 |
| columns | `TableColumn[]` | ''  | 必选，表格列配置|
| options | `TableOption` | `{ height: 'auto' }` | 可选，表格配置 |
| plugin  | `TablePlugin` | `{}` | 可选，表格插件配置 |

## TestTable 事件

| 参数 | 类型 | 说明 |
| :-: | :-: | - |
| pagingChange | `EventEmitter` | 可选，分页变化事件，返回当前页起始索引`pageStart`和当前页数量上限`pageLimit`，业务可以根据这两个值动态修改data，实现分页 |

# 接口 & 类型定义

TableOption
```ts
interface TableOption {

  /**
   * @default 'auto'
   */
  height: string;
}
```

TableColumn
```ts
export interface TableColumn {

  /**对应的数据字段 */
  field: string;

  /**列头显示的文案 */
  title: string;

  /**排序方法 */
  sortFn?: (...args: any[]) => number;
}
```

TablePlugin
```ts
interface TablePlugin {

  /**分页插件 */
  paging?: PagingPlugin;
}
```

pagingPlugin
```ts
export type PagingPlugin = false | {

  /**
   * @description 配置每页的数量上限
   * @default [20,50]
   */
  pageSize: number[];

  /**数据总数 */
  total: number;
};
```
