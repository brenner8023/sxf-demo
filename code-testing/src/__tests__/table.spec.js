import { ref } from '@vue/composition-api'
import { mount } from '@vue/test-utils'
import { TestTable } from '../table'
import TableBody from '../table/src/components/TableBody'

describe('Table', () => {
  const TableMount = options => mount(TestTable, options)
  const tableData = [
    {
      id: 1,
      age: 11,
    }, {
      id: 2,
      age: 22,
    }, {
      id: 3,
      age: 33,
    }, {
      id: 4,
      age: 33,
    }, {
      id: 5,
      age: 33,
    }, {
      id: 6,
      age: 33,
    }, {
      id: 7,
      age: 33,
    },
  ]
  const columns = [
    {
      field: 'id',
      title: 'id',
    }, {
      field: 'age',
      sortFn: (a, b) => a.age - b.age,
    },
  ]

  test('render', () => {
    const wrapper = TableMount()
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.vm.$destroy()
    }).not.toThrow()
  })

  test('custom slot', () => {
    const wrapper = mount({
      template: `
        <TestTable :data="data" :columns="columns">
          <template #action>
            <div id="action-aaa">编辑</div>
          </template>
        </TestTable>
      `,
      components: {
        TestTable,
      },
      setup () {
        return {
          data: ref([{ age: 1 }, { age: 2 }]),
          columns: ref([{ field: 'age', title: '年龄' }, { field: 'action', title: '操作' }]),
        }
      },
    })
    expect(wrapper.find('#action-aaa').exists()).toBe(true)
  })

  test('paging', async () => {
    const pagingChange = jest.fn()
    const wrapper = TableMount({
      propsData: {
        plugin: {
          paging: {
            pageSize: [5, 6],
            total: 99,
          },
        },
        data: tableData,
        columns,
      },
      listeners: {
        'paging-change': pagingChange,
      },
    })

    await wrapper.find('.sxf-paging-select').setValue(6)
    expect(pagingChange).toHaveBeenCalledTimes(1)

    await wrapper.find('.sxf-paging__input').setValue(2)
    await wrapper.find('.sxf-paging__input').trigger('change')
    expect(pagingChange).toHaveBeenCalledTimes(2)
  })

  test('table sort', async () => {
    const wrapper = TableMount({
      propsData: {
        data: tableData,
        columns,
      },
    })

    await wrapper.find('.sxf-table-sorter__desc').trigger('click')
    expect(wrapper.find('.sxf-table-sorter__desc.sxf-table-sorter__desc--selected').exists()).toBe(true)

    await wrapper.find('.sxf-table-sorter__asc').trigger('click')
    expect(wrapper.find('.sxf-table-sorter__asc.sxf-table-sorter__asc--selected').exists()).toBe(true)
  })

  test('table body', () => {
    const wrapper = mount(TableBody, {
      propsData: {
        isTableEmpty: true,
      },
    })

    expect(wrapper.find('.sxf-empty').exists()).toBe(true)
  })
})
