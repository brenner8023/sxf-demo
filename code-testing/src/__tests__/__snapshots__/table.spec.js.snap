// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render 1`] = `
"<div class=\\"sxf-table-wrap\\" style=\\"height: auto;\\">
  <table class=\\"sxf-table\\">
    <thead></thead>
    <tbody>
      <div class=\\"sxf-empty\\">空状态</div>
    </tbody>
  </table>
  <div class=\\"sxf-table-footer\\"></div>
</div>"
`;
